src/Backend/Haskell.idr

```diff
@@ -117,22 +347,386 @@ mutual
+||| given `TDef` depends on. Does not include the given `TDef` itself.
+||| @ env semantic environment, used for knowing what to name things
+dependencies : (env : Vect n HsType) -> TDef n -> List (m ** TNamed m)
+dependencies env td =
+  filter (\ (m ** t) => not $ heteroEq (def t) td)
+    (nubBy (\ (m ** t), (m' ** t') => heteroEqNamed t t') (go env td))
+  where
+  mutual
+    goMu : Vect n HsType -> Vect k (String, TDef (S n)) -> List (m ** TNamed m)
+    goMu env tds = concatMap (assert_total $ go ((makeType env (TMu tds))::env) . snd) tds
+
+    go : Vect n HsType -> TDef n -> List (m ** TNamed m)
+    go env T0 = []
+    go env T1 = []
+    go env (TSum xs) = concatMap (assert_total $ go env) xs
+    go env (TProd xs) = concatMap (assert_total $ go env) xs
+    go env (TVar v)    = []
+    go {n = n} env t@(TMu tds) = (n ** TName (nameMu tds) t) :: goMu env tds
+    go {n = n} env (TApp {k = k} t@(TName name td) xs) =
+        let envxs = map (makeType env) xs
+            -- dependencies of the actual td
+            depTd = case td of
+                     TMu tds => goMu envxs tds -- skip the TMu itself
+                     _       => go envxs td
+            -- function to fix up some unwanted entries
+            fixup = (\ x => case x of
+                              TApp f xs' => [] -- will be counted later
+                              TVar i     => [] -- will be a parameter anyway
+                              _          => [(n ** TName "" x)]) in
+        depTd ++ [(k**t)] ++ (concatMap (assert_total $ go env) xs) ++ (concatMap fixup xs)
+
+||| Given a TDef `td` and a Haskell term t of type [[ td ]] (where
+||| [[ td ]] is the interpretation of td as a type), constructs a
+||| Haskell Term of type `Builder`.
+||| @ td TDef to build term for
+||| @ t Haskell term of the TDef type the encoder should be applied to
+encode : (td : TDef n) -> (t : HsTerm) -> TermGen n HsTerm
+encode T0 t = pure (hsAbsurd t)
+encode T1 t = pure hsEmpty
+encode (TSum td) t = do
+  summands <- injectionInv td
+  pure $ HsCase t (map (\ (lhs, i, rhs) => (lhs, HsConcat [word i, rhs])) summands)
+ where injectionInv : Vect (2 + k) (TDef n) -> TermGen n (List (HsTerm, Int, HsTerm))
+       injectionInv [a, b] = do
+            [freshPV] <- freshVars 1 "z"
+            a' <- encode a freshPV
+            b' <- encode b freshPV
+            pure [ (hsLeft freshPV, 0, a')
+                 , (hsRight freshPV, 1, b')
+                 ]
+       injectionInv (a::b::c::tds) = do
+            [freshPV] <- freshVars 1 "z"
+            a' <- encode a freshPV
+            rest <- injectionInv (b::c::tds)
+            pure $ (hsLeft freshPV, 0, a') ::
+                     (map (\ (lhs, i, rhs) => (hsRight lhs, i + 1, rhs)) rest)
+encode (TProd {k} td) t = do
+   freshPVs <- freshVars (2 + k) "y"
+   factors <- mapWithIndexA (\ i, t' => assert_total $ encode (index i td) t') freshPVs
+   pure $ HsCase t [(HsTupC freshPVs, HsConcat (toList factors))]
```

Comment:
Assuming I understand your question correctly: no, this works because we translate a `TProd [a, b, c, ...]` into a flat tuple `(a', b', c', ...)` (with `a'` the translation of `a` etc). We never do nested tuples in the Haskell backend.

Reply:
In retrospect I think I totally misunderstood the line `HsCase t [(HsTupC freshPVs, HsConcat (toList factors))]` But it makes sense now. Thanks!